core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
-                        size += 1
-                        tailValues += value
-                        unblock = unblock <* producer.complete(())
+          def send(a: A): F[Either[Channel.Closed, Unit]] = {
+            // we track the outstanding blocked offers so we can distinguish closing from draining
+            // the very last blocked send, when closed, is responsible for triggering the sentinel
+            def permit[E](fe: F[E]): F[E] =
+              MonadCancel[F].uncancelable { poll =>
+                (leasesR.update(_ + 1) *> poll(fe)).guarantee {
+                  leasesR.updateAndGet(_ - 1).flatMap { leases =>
+                    if (leases <= 0) {
+                      closedR.get.flatMap { b =>
+                        if (b)
+                          q.offer(Sentinel)
+                            .start
+                            .void // we don't want to backpressure on processing the sentinel
+                        else
+                          Applicative[F].unit
                       }
-
-                      val toEmit = makeChunk(initValues, tailValues.result(), size)
-
-                      unblock.as(Pull.output(toEmit) >> consumeLoop)
                     } else {
-                      F.pure(
-                        if (closed) Pull.done
-                        else Pull.eval(waiting.get) >> consumeLoop
-                      )
+                      Applicative[F].unit
                     }
+                  }
                 }
-                .uncancelable
-            }
-          }.flatten
+              }
 
-        def notifyStream(waitForChanges: Option[Deferred[F, Unit]]) =
-          waitForChanges.traverse(_.complete(()))
+            isClosed.ifM(
+              LeftClosedF,
+              permit(isClosed.ifM(LeftClosedF, q.offer(a.asInstanceOf[AnyRef]).as(RightUnit)))
+            )
+          }
 
-        def waitOnBound(producer: Deferred[F, Unit], poll: Poll[F]) =
-          poll(producer.get).onCancel {
-            state.update { s =>
-              s.copy(producers = s.producers.filter(_._2 ne producer))
+          def trySend(a: A): F[Either[Channel.Closed, Boolean]] =
+            isClosed.flatMap { b =>
+              if (b)
+                LeftClosedF.asInstanceOf[F[Either[Channel.Closed, Boolean]]]
+              else
+                q.tryOffer(a.asInstanceOf[AnyRef]).map(_.asRight[Channel.Closed])
             }
-          }
 
-        def signalClosure = closedGate.complete(())
+          val stream: Stream[F, A] = {
+            val takeN: F[Chunk[A]] =
+              q.tryTakeN(None).flatMap {
+                case Nil =>
+                  // if we land here, it either means we're consuming faster than producing
+                  // or it means we're actually closed and we need to shut down
+                  // this is the unhappy path either way
+
+                  val fallback = q.take.flatMap { a =>
```

Comment:
Imagine this scenario: queue empty for a while, then a bunch of producers enqueue at once. The old channel has a chance of  chunking up all the new updates up, this code doesn't. It will probably get picked up in the next iteration, but it doesn't seem hard to fix either

Reply:
So in this implementation, the following will happen in `stream` deterministically:

1. A `singleton` chunk will be emitted containing the first-ish produced element
2. A `seq` chunk will be emitted containing all of the remainder

In the original implementation, it was possible to get two or even just one `seq` chunk, but only if the consumer *repeatedly lost the race*. In other words, this is a pretty fundamental tradeoff between latency and throughput. The new implementation actually has better latency because it will deterministically give you an element *as soon* as it is available, whereas the old implementation could do that but also had a chance to just spin for a little while and then give you a bigger chunk.

Ultimately, singletons are the base mode of operation for `Channel`. When producers and consumers are at the same flow rate, every chunk will be `singleton`. This is good because it means everyone is keeping up! `seq` chunks only happen when the consumer is slower than the aggregate producer flow rate, resulting in trading off latency for throughput in an attempt to catch up. This can be viewed as a degraded case. Valuable in an elastic system, but still degraded.

Reply:
>  but it doesn't seem hard to fix either
I take this back because I realised there isn't a `takeAvailable` on Queue. That being said, not really agree with

>  resulting in trading off latency for throughput in an attempt to catch up

because it's not like the old implementation adds linger time in order to catch up, it uses the latency it's already wasted to potentially gain better throughput

> e new implementation actually has better latency because it will deterministically give you an element as soon as it is available

Isn't that the same as the old channel, i.e. a `Deferred[Unit]` triggers, except strictly worse because it "discards" additional elements that might have been added in the mean time?

That being said, it's probably not worth changing since there's no `takeAvailable`

Reply:
> but sometimes the consumer would lose the race since the whole state was synchronized, meaning that it would spin again trying to grab its chunk

Ah I see what you mean now